### PR TITLE
Refactor: APPX-87: Fix input passing

### PR DESCRIPTION
**Pull request by martini-source**
Changes:
- use --kind flag for serviceDiscovery 
- use first argument passed by args as serviceName
- require --kind flag to be set

### CODE REVIEW

File: internal/util/argParser.go [hunk 1]
```diff
@@ -122,6 +99,51 @@ func CheckForMissingValues(arguments []string) error {
...
+func FilterServiceKind(input []Input) (string, error) {
+	serviceKindInput := filterInput(input, func(input Input) bool {
+		return input.ParameterHierarchy[0] == K8S_SERVICE_KIND
+	})
+	if len(serviceKindInput) != 1 {
+		return "", fmt.Errorf("ServiceKind is missing")
+	} else {
+		return serviceKindInput[0].Value, nil
+	}
+
```
Comment (davidgubler) [inline internal/util/argParser.go:111]: empty line

**Comment by davidgubler** [comment]
I didn't dive into the appcat-cli codebase very deeply but from what I can see the changes look reasonable.

File: internal/util/argParser.go [hunk 2]
```diff
@@ -122,6 +99,50 @@ func CheckForMissingValues(arguments []string) error {
...
+func filterInput(slice []Input, predicate func(Input) bool) []Input {
+	var result []Input
+	for _, item := range slice {
+		if predicate(item) {
+			result = append(result, item)
+		}
+	}
+	return result
+}
```
Comment (mhutter) [inline internal/util/argParser.go:121]: Did you reimplement [slices.IndexFunc](https://pkg.go.dev/slices@go1.21.0#IndexFunc)? :D

File: main.go [hunk 3]
```diff
@@ -56,36 +56,44 @@ func main() {
 // If during execution, an unrecoverable error occurs (usually due to a bug),
 // an error message is logged and the program will panic.
 func Main(apps applications.AppMap, args []string, out io.Writer) int {
-	if len(args) < 2 {
+	if len(args) < 4 {
```
Comment (mhutter) [inline main.go:59]: I think if you'd write `--kind=Foobar` it's only 3, right?

File: main.go [hunk 4]
```diff
@@ -56,36 +56,44 @@ func main() {
...
+	app, ok := apps[serviceKind]
 	if !ok {
-		logrus.Errorf("service type '%s' is not supported", parsedType)
+		logrus.Errorf("service type '%s' is not supported", serviceKind)
 		printUsage(args[0], apps)
 		return 1
 	}
-
-	var err error
+	////appcat-cli serviceKind
```
Comment (mhutter) [inline main.go:89]: leftover

**Review by mhutter — CHANGES_REQUESTED** [review]
See notes, rest LGTM :+1:

File: internal/util/argParser.go (hunk 2 above)
Comment (martini-source) [reply internal/util/argParser.go:121]: Not exactly, but yes ':)

File: main.go (hunk 3 above)
Comment (martini-source) [reply main.go:59]: True, I changed it to 3, should be okay since there is a check later on if all key have their values and if --kind is present.

**Review by mhutter — APPROVED** [review]

**Comment by mhutter** [comment]
We'll have to replace `golang.org/x/exp/slices` with the stdlib one's once we upgraded to Go 1.21

**Pull request merged**